File: frontend/src/types/trading.ts
// types/trading.ts

export type LoadingState = "idle" | "loading" | "success" | "error";

export interface TradingRoute {
  rank?: number; // Optional for API responses
  item_type_id: number;
  item_name: string;
  // API fields
  buy_system_id?: number;
  buy_system_name?: string;
  buy_station_id?: number;
  buy_station_name?: string;
  sell_system_id?: number;
  sell_system_name?: string;
  sell_station_id?: number;
  sell_station_name?: string;
  buy_security_status?: number; // Security status of buy system (0.0 - 1.0)
  sell_security_status?: number; // Security status of sell system (0.0 - 1.0)
  min_route_security_status?: number; // Minimum security of all systems on route
  // Mock data fields (for backward compatibility)
  origin_system_id?: number;
  origin_system_name?: string;
  destination_system_id?: number;
  destination_system_name?: string;
  quantity: number;
  buy_price: number;
  sell_price: number;
  total_investment?: number;
  total_revenue?: number;
  profit?: number; // Mock data
  total_profit?: number; // API field
  spread_percent: number;
  travel_time_seconds: number;
  round_trip_seconds: number;
  isk_per_hour: number;
  jumps?: number; // API field
  item_volume?: number; // API field
  profit_per_unit?: number; // API field
  // Multi-tour fields
  number_of_tours?: number;
  profit_per_tour?: number;
  total_time_minutes?: number;
  // Navigation Skills fields
  base_travel_time_seconds?: number; // Travel time without navigation skills
  skilled_travel_time_seconds?: number; // Travel time with navigation skills applied
  base_isk_per_hour?: number; // ISK/h without navigation skills
  time_improvement_percent?: number; // Percentage improvement from skills
  // Cargo fields
  cargo_used?: number; // m³ actually used
  cargo_capacity?: number; // Total effective capacity (with skills + fitting)
  cargo_utilization?: number; // Percentage 0-100
  base_cargo_capacity?: number; // Base capacity without skills
  skill_bonus_percent?: number; // Total skill bonus %
  fitting_bonus_m3?: number; // Fitting modules bonus (absolute m³)
  // Fee fields
  gross_profit?: number; // Revenue - Investment (before fees)
  gross_margin_percent?: number; // (gross_profit / total_investment) * 100
  sales_tax?: number; // Sales tax fee
  broker_fees?: number; // Broker fees
  estimated_relist_fee?: number; // Estimated relist fee
  total_fees?: number; // Sum of all fees
  net_profit?: number; // Gross profit - total fees
  net_profit_percent?: number; // Net margin percentage
  // Volume & Liquidity fields (Issue #53)
  volume_metrics?: VolumeMetrics; // Market volume and liquidity data
  liquidation_days?: number; // Estimated days to sell inventory
  daily_profit?: number; // Profit per day (net_profit / liquidation_days)
}

export interface VolumeMetrics {
  type_id: number;
  region_id: number;
  daily_volume_avg: number; // 30-day average daily volume
  daily_isk_turnover: number; // Average daily ISK traded (volume × avg_price)
  liquidity_score: number; // 0-100 score based on volume stability
  data_days: number; // Number of days of historical data available
}

export interface Region {
  id: number;
  name: string;
}

export interface Ship {
  item_id: number;
  type_id: number;
  name: string;
  /** Base hull cargo (without skills/modules). */
  cargo_capacity: number;
  /** Effective cargo (hull + skills + fitted modules); same value the optimizer
   *  uses. Absent when the ship has no cargo-expander fitting — then the base
   *  `cargo_capacity` ("Basis") is the correct value. */
  effective_cargo_capacity?: number;
  /** True when the backend's fitting enrichment ERRORED for this ship (ESI/
   *  network), so the effective volume is genuinely unknown. The UI shows a
   *  visible "unknown" marker rather than passing off the base hull as fitted. */
  effective_cargo_unavailable?: boolean;
}

export interface TradingFilters {
  minSpread: number;
  minProfit: number;
  maxTravelTime: number;
  allowHighSec: boolean;
  allowLowSec: boolean;
  allowNullSec: boolean;
  // Volume filters (Issue #53)
  minDailyVolume?: number; // Minimum daily volume (items/day)
  maxLiquidationDays?: number; // Maximum liquidation time (days)
  includeVolumeMetrics?: boolean; // Whether to include volume metrics
}

export interface ItemSearchResult {
  type_id: number;
  name: string;
  group_name: string;
}

// Multi-Hub Comparison (Issue #43)

export interface CompetitionInfo {
  changes_per_hour: number;
  source: "live" | "baseline" | string;
}

export interface SkillsApplied {
  applied: boolean;
  accounting: number;
  broker_relations: number;
  sales_tax_rate: number;
  broker_fee_rate: number;
}

export interface HubRow {
  region_id: number;
  region_name: string;
  hub_name: string;
  system_id: number;
  tier: string;
  has_data: boolean;
  buy_price: number;
  sell_price: number;
  spread_percent: number;
  net_margin_percent: number;
  net_profit_per_unit: number;
  daily_volume: number;
  liquidity_score: number;
  competition: CompetitionInfo;
}

export interface HubComparisonResult {
  type_id: number;
  item_name: string;
  best_hub_region_id: number;
  skills_applied: SkillsApplied;
  hubs: HubRow[];
}

// ROI Calculator & Capital Allocation Optimizer (Issue #44)

export interface PortfolioRequest {
  region_id: number;
  ship_type_id: number;
  capital: number;
  time_budget_min: number;
  liquidity_cap_pct: number;
  max_item_pct: number;
  sec_zones: string[]; // e.g. ["high", "low", "null"]
  /** Selected ship instance's effective cargo (m³). When set, the optimizer
   *  uses this exact value instead of recomputing per ship type. Omitted when
   *  the instance's effective cargo is unavailable/unknown (backend falls back
   *  to the per-type recompute). */
  cargo_capacity?: number;
}

export interface PortfolioItem {
  type_id: number;
  name: string;
  capital_used: number;
  units: number;
  trips_per_day: number;
  daily_profit: number;
  roi_percent: number;
  /** Daily profit per hour of haul time. Items are pre-sorted by this metric. */
  isk_per_hour: number;
  // Buy/sell location (station → station haul)
  buy_system_name: string;
  buy_station_name: string;
  buy_station_id: number;
  sell_system_name: string;
  sell_station_name: string;
  sell_station_id: number;
}

export interface PortfolioResult {
  items: PortfolioItem[];
  total_capital_used: number;
  total_daily_profit: number;
  time_used_min: number;
  diversification_score: number;
  skills_applied: SkillsApplied;
}

// Neighborhood Hauling Routes (Issue #45)

export type SecurityRisk = "safe" | "caution" | "danger";

export interface HaulingRequest {
  origin_region_id: number; // 0 ⇒ backend uses character's current region
  ship_type_id: number;
  capital: number;
  avoid_low_sec: boolean;
  max_routes: number;
  /** Selected ship instance's effective cargo (m³). When set, the backend uses
   *  this exact value instead of recomputing per ship type. Omitted when the
   *  instance's effective cargo is unavailable/unknown (backend falls back to
   *  the per-type recompute). */
  cargo_capacity?: number;
}

export interface HaulingItem {
  type_id: number;
  name: string;
  quantity: number;
  buy_price: number;
  sell_price: number;
  unit_volume: number;
  profit: number;
  profit_per_m3: number;
}

export interface HaulingRoute {
  buy_system_name: string;
  buy_station_name: string;
  buy_station_id: number;
  sell_system_name: string;
  sell_station_name: string;
  sell_station_id: number;
  jumps: number;
  travel_time_min: number;
  security_risk: SecurityRisk;
  total_profit: number;
  total_capital: number;
  total_volume: number;
  isk_per_hour: number;
  items: HaulingItem[];
}

export interface HaulingResponse {
  origin_region_id: number;
  regions_scanned: number[];
  routes: HaulingRoute[]; // pre-sorted by isk_per_hour desc
  skills_applied: SkillsApplied;
}

export interface InventorySellRequest {
  type_id: number;
  quantity: number;
  buy_price_per_unit: number;
  region_id: number;
  min_profit_per_unit: number;
  security_filter: string; // "all", "highsec", "highlow"
}

export interface InventorySellRoute {
  sell_station_id: number;
  sell_station_name: string;
  sell_system_id: number;
  sell_system_name: string;
  sell_security_status: number;
  buy_order_price: number;
  tax_rate: number;
  net_price_per_unit: number;
  profit_per_unit: number;
  available_quantity: number;
  total_profit: number;
  route_jumps: number;
  route_system_ids: number[];
  min_route_security_status: number;
}

// Sell-from-Assets (Issue #107)

export interface AssetItem {
  type_id: number;
  name: string;
  quantity: number;
  location_id: number;
  location_name: string;
  system_id: number;
  region_id: number;
  marketable: boolean;
}

export interface AssetsResponse {
  assets: AssetItem[];
  count: number;
  /** ESI cache-expiry (RFC 3339). Until this point ESI serves the same
   *  snapshot, so a refresh before then returns identical data. Absent when
   *  ESI didn't return a parseable Expires header. */
  cache_expires_at?: string;
}

export interface SellOptionsRequest {
  type_id: number;
  location_id: number;
  quantity: number;
  avoid_low_sec: boolean;
}

export type SellOptionScope = "hub" | "current_region";

export interface SellOption {
  scope: SellOptionScope;
  region_id: number;
  region_name: string;
  station_id: number;
  station_name: string;
  system_name: string;
  buy_price: number;
  unit_net: number;
  total_net: number;
  /** Net ISK per hour of travel. 0 for local sales (no travel) — they rank
   *  ahead of any remote sale by `total_net`. */
  isk_per_hour: number;
  jumps: number;
  travel_time_min: number;
  security_risk: SecurityRisk;
  has_data: boolean;
}

export interface SellOptionsResponse {
  type_id: number;
  name: string;
  quantity: number;
  origin_system_id: number;
  best: SellOption | null;
  options: SellOption[]; // pre-sorted by total_net desc
  skills_applied: SkillsApplied;
  /** Set when the empty options list deserves an explanation. Current values:
   *  "origin_in_player_structure" — the item sits in a citadel; SDE can't
   *  resolve its system, so no route is possible. */
  not_routable_reason?: string;
}

// Mining Ore Ranking (ore ISK/hour calculator)

export interface OreRankingRequest {
  region_id: number;
  sec_band: string; // "high" | "low" | "null"
}

export interface SellLocation {
  station_name?: string;
  system_name?: string;
  is_structure: boolean;
  location_id?: number;
}

export interface RefineMaterial {
  material_type_id: number;
  material_name: string;
  effective_qty: number;
  buy_price: number;
  sell: SellLocation;
}

export interface OreRankRow {
  ore_type_id: number;
  ore_name: string;
  mining_m3_per_hour: number;
  raw_isk_per_hour: number;
  refine_isk_per_hour: number;
  raw_net_per_m3: number;
  refine_net_per_m3: number;
  best: string; // "raw" | "refine"
  delta_isk_per_hour: number;
  best_station_id?: number;
  best_station_tax: number;
  best_station_name?: string;
  best_station_system?: string;
  raw_sell?: SellLocation;
  materials?: RefineMaterial[];
  hull_yield_multiplier?: number;
  crystal_multiplier?: number;
  is_estimate?: boolean;
  estimate_reason?: string;
  effective_isk_per_hour?: number;
  load_volume_m3?: number;
  fill_minutes?: number;
  cycle_minutes?: number;
  route_jumps?: number;
  sell_system_name?: string;
}

export interface OreRankingResponse {
  region_id: number;
  sec_band: string;
  no_mining_setup: boolean;
  rows: OreRankRow[];
}

File: backend/internal/models/trading.go
// Package models provides data structures for trading operations
package models

import "time"

// TradingRoute represents a profitable trading route
type TradingRoute struct {
	ItemTypeID             int     `json:"item_type_id"`
	ItemName               string  `json:"item_name"`
	BuySystemID            int64   `json:"buy_system_id"`
	BuySystemName          string  `json:"buy_system_name"`
	BuyStationID           int64   `json:"buy_station_id"`
	BuyStationName         string  `json:"buy_station_name"`
	BuyPrice               float64 `json:"buy_price"`
	SellSystemID           int64   `json:"sell_system_id"`
	SellSystemName         string  `json:"sell_system_name"`
	SellStationID          int64   `json:"sell_station_id"`
	SellStationName        string  `json:"sell_station_name"`
	SellPrice              float64 `json:"sell_price"`
	BuySecurityStatus      float64 `json:"buy_security_status"`
	SellSecurityStatus     float64 `json:"sell_security_status"`
	MinRouteSecurityStatus float64 `json:"min_route_security_status"` // Minimum security of all systems on route
	Quantity               int     `json:"quantity"`
	ProfitPerUnit          float64 `json:"profit_per_unit"`
	TotalProfit            float64 `json:"total_profit"`
	SpreadPercent          float64 `json:"spread_percent"`
	TravelTimeSeconds      float64 `json:"travel_time_seconds"`
	RoundTripSeconds       float64 `json:"round_trip_seconds"`
	ISKPerHour             float64 `json:"isk_per_hour"`
	Jumps                  int     `json:"jumps"`
	ItemVolume             float64 `json:"item_volume"`
	// Multi-tour fields
	NumberOfTours    int     `json:"number_of_tours"`
	ProfitPerTour    float64 `json:"profit_per_tour"`
	TotalTimeMinutes float64 `json:"total_time_minutes"`
	// Navigation Skills fields
	BaseTravelTimeSeconds    float64 `json:"base_travel_time_seconds"`    // Travel time without navigation skills
	SkilledTravelTimeSeconds float64 `json:"skilled_travel_time_seconds"` // Travel time with navigation skills applied
	BaseISKPerHour           float64 `json:"base_isk_per_hour"`           // ISK/h without navigation skills
	TimeImprovementPercent   float64 `json:"time_improvement_percent"`    // Percentage improvement from skills
}

// RouteCalculationRequest represents the request to calculate trading routes
type RouteCalculationRequest struct {
	RegionID      int     `json:"region_id"`
	ShipTypeID    int     `json:"ship_type_id"`
	CargoCapacity float64 `json:"cargo_capacity,omitempty"`
}

// RouteCalculationResponse represents the response with calculated routes
type RouteCalculationResponse struct {
	RegionID          int            `json:"region_id"`
	RegionName        string         `json:"region_name"`
	ShipTypeID        int            `json:"ship_type_id"`
	ShipName          string         `json:"ship_name"`
	CargoCapacity     float64        `json:"cargo_capacity"`
	CalculationTimeMS int64          `json:"calculation_time_ms"`
	Routes            []TradingRoute `json:"routes"`
	Warning           string         `json:"warning,omitempty"`
}

// ItemPair represents a profitable buy/sell opportunity for an item
type ItemPair struct {
	TypeID            int     `json:"type_id"`
	ItemName          string  `json:"item_name"`
	ItemVolume        float64 `json:"item_volume"`
	BuyStationID      int64   `json:"buy_station_id"`
	BuySystemID       int64   `json:"buy_system_id"`
	BuyPrice          float64 `json:"buy_price"`
	SellStationID     int64   `json:"sell_station_id"`
	SellSystemID      int64   `json:"sell_system_id"`
	SellPrice         float64 `json:"sell_price"`
	SpreadPercent     float64 `json:"spread_percent"`
	AvailableVolumeM3 float64 `json:"available_volume_m3"` // Total m³ available from sell orders
	AvailableQuantity int     `json:"available_quantity"`  // Total items available
}

// CharacterLocation represents character location information
type CharacterLocation struct {
	SolarSystemID   int64   `json:"solar_system_id"`
	SolarSystemName string  `json:"solar_system_name"`
	RegionID        int64   `json:"region_id"`
	RegionName      string  `json:"region_name"`
	StationID       *int64  `json:"station_id,omitempty"`
	StationName     *string `json:"station_name,omitempty"`
	StructureID     *int64  `json:"structure_id,omitempty"`
}

// CharacterShip represents character ship information
type CharacterShip struct {
	ShipTypeID    int64   `json:"ship_type_id"`
	ShipName      string  `json:"ship_name"`
	ShipItemID    int64   `json:"ship_item_id"`
	ShipTypeName  string  `json:"ship_type_name"`
	CargoCapacity float64 `json:"cargo_capacity"`
}

// CharacterAssetShip represents a ship in character assets
type CharacterAssetShip struct {
	ItemID        int64   `json:"item_id"`
	TypeID        int64   `json:"type_id"`
	TypeName      string  `json:"type_name"`
	LocationID    int64   `json:"location_id"`
	LocationName  string  `json:"location_name"`
	LocationFlag  string  `json:"location_flag"`
	CargoCapacity float64 `json:"cargo_capacity"`
	IsSingleton   bool    `json:"is_singleton"`
}

// CharacterShipsResponse represents the response for character ships
type CharacterShipsResponse struct {
	Ships []CharacterAssetShip `json:"ships"`
	Count int                  `json:"count"`
}

// CachedData represents cached market data
type CachedData struct {
	Data      interface{}
	ExpiresAt time.Time
}

// Region represents an EVE Online region
type Region struct {
	ID   int64  `json:"id"`
	Name string `json:"name"`
}

// RegionsResponse represents the response for regions list
type RegionsResponse struct {
	Regions []Region `json:"regions"`
	Count   int      `json:"count"`
}

// ItemSearchResult represents a search result for items
type ItemSearchResult struct {
	TypeID    int    `json:"type_id"`
	Name      string `json:"name"`
	GroupName string `json:"group_name"`
}

// InventorySellRequest represents a request to find profitable sell routes for inventory
type InventorySellRequest struct {
	TypeID           int     `json:"type_id"`
	Quantity         int     `json:"quantity"`
	BuyPricePerUnit  float64 `json:"buy_price_per_unit"`
	RegionID         int     `json:"region_id"`
	MinProfitPerUnit float64 `json:"min_profit_per_unit"`
	SecurityFilter   string  `json:"security_filter"` // "all", "highsec", "highlow"
}

// Validate validates the inventory sell request parameters
func (r *InventorySellRequest) Validate() error {
	if r.TypeID <= 0 {
		return &ValidationError{Field: "type_id", Message: "Invalid type_id"}
	}
	if r.Quantity <= 0 {
		return &ValidationError{Field: "quantity", Message: "Invalid quantity"}
	}
	if r.BuyPricePerUnit <= 0 {
		return &ValidationError{Field: "buy_price_per_unit", Message: "Invalid buy_price_per_unit"}
	}
	if r.RegionID <= 0 {
		return &ValidationError{Field: "region_id", Message: "Invalid region_id"}
	}
	return nil
}

// ValidationError represents a validation error
type ValidationError struct {
	Field   string
	Message string
}

func (e *ValidationError) Error() string {
	return e.Message
}

// ShipType represents a ship with its navigation characteristics
type ShipType struct {
	TypeID        int     // EVE Online Type ID
	Name          string  // Ship name
	BaseWarpSpeed float64 // Base warp speed in AU/s (typically 3.0 for haulers)
	BaseAlignTime float64 // Base align time in seconds
	CargoCapacity float64 // Cargo capacity in m³
}

// CommonHaulers defines commonly used hauler ships with their base stats
// Warp speeds are in AU/s, align times in seconds
var CommonHaulers = map[int]ShipType{
	648: { // Iteron Mark V (Gallente)
		TypeID:        648,
		Name:          "Iteron Mark V",
		BaseWarpSpeed: 3.0,
		BaseAlignTime: 8.5,
		CargoCapacity: 63000,
	},
	649: { // Badger (Caldari)
		TypeID:        649,
		Name:          "Badger",
		BaseWarpSpeed: 3.0,
		BaseAlignTime: 7.2,
		CargoCapacity: 50000,
	},
	650: { // Nereus (Gallente)
		TypeID:        650,
		Name:          "Nereus",
		BaseWarpSpeed: 3.0,
		BaseAlignTime: 6.8,
		CargoCapacity: 45000,
	},
	651: { // Wreathe (Minmatar)
		TypeID:        651,
		Name:          "Wreathe",
		BaseWarpSpeed: 3.0,
		BaseAlignTime: 8.0,
		CargoCapacity: 48000,
	},
	// Default hauler (used as fallback)
	0: {
		TypeID:        0,
		Name:          "Generic Hauler",
		BaseWarpSpeed: 3.0,
		BaseAlignTime: 8.0,
		CargoCapacity: 50000,
	},
}

// GetShipType returns ship type by ID, or default hauler if not found
func GetShipType(typeID int) ShipType {
	if ship, ok := CommonHaulers[typeID]; ok {
		return ship
	}
	return CommonHaulers[0] // Return default hauler
}

// InventorySellRoute represents a profitable sell opportunity for inventory
type InventorySellRoute struct {
	SellStationID          int64   `json:"sell_station_id"`
	SellStationName        string  `json:"sell_station_name"`
	SellSystemID           int64   `json:"sell_system_id"`
	SellSystemName         string  `json:"sell_system_name"`
	SellSecurityStatus     float64 `json:"sell_security_status"`
	BuyOrderPrice          float64 `json:"buy_order_price"`
	TaxRate                float64 `json:"tax_rate"`
	NetPricePerUnit        float64 `json:"net_price_per_unit"`
	ProfitPerUnit          float64 `json:"profit_per_unit"`
	AvailableQuantity      int     `json:"available_quantity"`
	TotalProfit            float64 `json:"total_profit"`
	RouteJumps             int     `json:"route_jumps"`
	RouteSystemIDs         []int64 `json:"route_system_ids"`
	MinRouteSecurityStatus float64 `json:"min_route_security_status"`
}
